HAK-26: Documentation for channels API after refactor (#2883)
tloncorp/landscape-apps: docs/channels.md

| `/x/[kind]/[ship]/[name]/posts/...` | [channel-posts](https://github.com/tloncorp/landscape-apps/blob/hm/hackweek/desk/mar/channel/posts.hoon) | Get posts from a channel. This can be called with a large variety of arguments, found [here](https://github.com/tloncorp/landscape-apps/blob/hm/hackweek/desk/app/channels.hoon#L1044). Returns [paged-posts](https://github.com/tloncorp/landscape-apps/blob/hm/hackweek/desk/sur/channel.hoon#L423) | <code>.^(scan:c %gx /=channels=/chat/~sampel-palnet/mychat/posts/newest/10/post</code>
| `/u/[kind]/[ship]/[name]` | loob | Check the existence of a channel | <code>.^(? %gu /=channels=/chat/~sampel-palnet/mychat/loob)</code>

## Pokes
All pokes here have the [channel-action](https://github.com/tloncorp/landscape-apps/blob/hm/hackweek/desk/mar/channel/action.hoon) mark.

### Channels
| Description | Example
|-------------|--------

> | Create a channel. This also updates the metadata in the `%groups` agent. See [$create-channel](https://github.com/tloncorp/landscape-apps/blob/hm/hackweek/desk/sur/channel.hoon#L245) | <pre>=create-channel :\*<br>  %chat<br>  %mychat<br>  [our %mygroup]<br>  title='My chat channel'<br>  description='Chat about things'<br>  readers=\~<br>  writers=\~<br>==<br><br>:channels &channel-action [%create create-channel]</pre>

Reviewer: Wondering if these would benefit from having a first column that super briefly describes the shape or tag of the action.

Maybe like `[%create *]`, `[%channel * %unwatch *]`, `[%channel * %post %reply * %del-react *]`,  
or even terser like `%create`, `%unwatch`, `%del-react` respectively, and then just mentioning the `[%channel *]` case and such, as pointers to the following section.

Unsure if that would be better or not, but currently feels like there's some indexing/skimmability stuff missing here. The descriptions themselves are a bit too verbose for that.

| Allow writing to a channel for certain roles. | <pre>=nest [%chat our %mychat]<br>=action [%add-writers (sy ~[%role1 %role2])]<br><br>:channels &channel-action [%channel nest action]</pre>
| Disallow writing to a diary for a certain set of roles. | <pre>=nest [%chat our %mychat]<br>=action [%del-writers (sy ~[%role1 %role2])]<br><br>:channels &channel-action [channel nest action]</pre>
| Pin channels | <pre>=pins ~[[%chat ~pinner %mychat]]<br><br>:channels &channel-action [%pin pins]</pre>

### Unreads
| Description | Example
|-------------|--------
| Mark a channel as read. | <pre>=nest [%chat our %mychat]<br>=action [%read ~]<br><br>:channels &channel-action [%channel nest action]</pre>
| Mark a channel as read up to a certain time. | <pre>=nest [%chat our %mychat]<br>=action [%read-at now]<br><br>:channels &channel-action [%channel nest action]</pre>
| Watch a channel for unreads. This affects notifications on the frontend. | <pre>=nest [%chat our %mychat]<br>=action [%watch ~]<br><br>:channels &channel-action [%channel nest action]</pre>

> | Stop watching a channel for unreads. This affects notifications on the frontend. | <pre>=nest [%chat our %mychat]<br>=action [%unwatch ~]<br><br>:channels &channel-action [%channel nest action]</pre>

Author: Just found out that these are currently unused, so I added a note about that.

> # Channels

Reviewer 2: We probably should have some kind of introductory paragraph about what channels is, what kinds of posts we have and how those are displayed. @jamesacklin wrote some stuff up in the readme, maybe can help with this. This can wait until later though just leaving a note here.

```

Join a channel. In this example, ~sampel-palnet is the group host and ~palnet-sampel is the channel host.
###### %leave
```hoon
=nest [%chat ~palnet-sampel %mychat]
=action [%leave ~]

:channels &channel-action [%channel nest action]
```

Leave a channel
###### %read
```hoon
=nest [%chat our %mychat]
=action [%read ~]

:channels &channel-action [%channel nest action]
```

Mark a channel as read.
###### %read-at
```hoon
=nest [%chat our %mychat]
=action [%read-at now]

:channels &channel-action [%channel nest action]
```

Mark a channel as read up to a certain time.
###### %watch
```hoon
=nest [%chat our %mychat]
=action [%watch ~]

:channels &channel-action [%channel nest action]
```

Watch a channel for unreads.
###### %unwatch
```hoon
=nest [%chat our %mychat]
=action [%unwatch ~]

:channels &channel-action [%channel nest action]
```

Stop watching a channel for unreads.
###### %post %add
```hoon
=nest [%chat our %mychat]
=memo :*
  ~[[%inline ~['hello world!']]]
  author=our
  sent=now
==
=kind-data [%chat ~]
=essay [memo kind-data]
=post-action [%add essay]
=action [%post post-action]

:channels &channel-action [%channel nest action]
```

> 

Reviewer 2: A marker here for where commands start and maybe an explainer about what that means semantically.

`%listing`: a traditional HTML list, ul and ol
`%code`: a block of code

[listing](#listing)
### listing
```hoon
+$  listing
  $%  [%list p=?(%ordered %unordered %tasklist) q=(list listing) r=(list inline)]
      [%item p=(list inline)]
  ==
```

HTML-style list. Recursively nesting

[inline](#inline)
### inline
```hoon
+$  inline  $+  channel-inline
  $@  @t
  $%  [%italics p=(list inline)]
      [%bold p=(list inline)]
      [%strike p=(list inline)]
      [%blockquote p=(list inline)]
      [%inline-code p=cord]
      [%code p=cord]
      [%ship p=ship]
      [%block p=@ud q=cord]
      [%tag p=cord]
      [%link p=cord q=cord]
      [%task p=?(%.y %.n) q=(list inline)]
      [%break ~]
  ==
```

Chunk of post content that can live inside of a paragraph

`@t`: plain text
`%italics`: italic text
`%bold`: bold text
`%strike`: strikethrough text
`%inline-code`: code formatting for small snippets
`%blockquote`: blockquote surrounded content
`%block`: link/reference to blocks
`%code`: code formatting for large snippets
`%tag`: tag gets special signifier
`%link`: link to a URL with a face
`%break`: line break
### kind-data
```hoon
+$  kind-data

>   $%  [%channel title=@t image=@t]

Reviewer 2: ```suggestion
  $%  [%diary title=@t image=@t]
```

### sort
```hoon
+$  sort  $~(%time ?(%alpha %time %arranged))
```

The persisted sort format for a channel
### arranged-posts
```hoon
+$  arranged-posts  (unit (list time))
```

Manually arranged posts. If null, use ordinary sorting.
### unreads
```hoon
+$  unreads  (map nest unread)
```

Unread info for channels. Map of [nest](#nest) to [unread](#unread)

### unread
```hoon
+$  unread   [last=time count=@ud read-id=(unit time)]
```

Unread info for a channel. 

`last`: last read time
`count`: number of unread messages
`read-id`: ID of last read message
### r-channels
```hoon
+$  r-channels  [=nest =r-channel]
```
Response (subscriber to client communication) for a channels

[nest](#nest) \| [r-channel](#r-channel)
### r-channel
```hoon
+$  r-channel
  $%  [%posts =notes]
      [%post id=id-post =r-post]
      [%set-order order=arranged-posts]
      [%view =view]
      [%sort =sort]
      [%perm =perm]
    ::
      [%create =perm]
      [%join group=flag:g]
      [%leave ~]
    ::
      [%read ~]
      [%read-at =time]
      [%watch ~]
      [%unwatch ~]
  ==
```

Response (subscriber to client communication) for a channel.
### posts
```hoon

> +$  notes  ((mop id-post (unit note)) lte)

Reviewer 2: ```suggestion
+$  posts  ((mop id-post (unit post)) lte)
```

      [%view =view]
      [%sort =sort]
      [%perm =perm]
    ::
      [%create =perm]
      [%join group=flag:g]
      [%leave ~]
    ::
      [%read ~]
      [%read-at =time]
      [%watch ~]
      [%unwatch ~]
  ==
```

Response (subscriber to client communication) for a channel.

### channels
```hoon
+$  channels  (map nest channel)
```

A map of [nest](#nest) to [channel](#channel)

### channel
```hoon
++  channel
  |^  ,[global local]
  +$  global
    $:  posts=posts
        order=arranged-posts
        =view
        =sort
        =perm
    ==
  ::
  +$  local
    $:  =net
        =remark
    ==
  --
```

Posts and view/sort/order info for a channel

[posts](#posts) \| [arranged-posts](#arranged-posts) \| [view](#view) \| [sort](#sort) \| [perm](#perm)

### perm
```hoon
+$  perm
  $:  writers=(set sect:g)
      group=flag:g
  ==
```

Permissions for a channel

[sect:groups](#sectgroups) \| [flag:groups](#flaggroups)
### v-channel
Internal representation of a channel. Contains syncing information and revision numbers.

[channel](#channel)

### nest
```hoon
+$  nest  [=kind =ship name=term]
```

Uniquely identifies a channel.

[kind](#kind)

### kind
```hoon

> +$  han  ?(%notebook %gallery %chat)

Reviewer 2: ```suggestion
+$  kind  ?(%notebook %gallery %chat)
```

.^(channels:c %gx /=channels=/channels/channel-channels)
```

Get the channels our ship is in and their contents. Returns `channels`

[channels](#channels-1)
#### /unreads
```hoon
=c -build-file /=groups=/sur/channel/hoon
.^(unreads:c %gx /=channels=/unreads/channel-unreads)
```

Get unread information for the channels our ship is in. Returns `unreads`

[unreads](#unreads-2)
#### /init
```hoon
=c -build-file /=groups=/sur/channel/hoon
.^([unreads:c channels:c] %gx /=channels=/init/noun)
```

Combination of `/channels` and `/unreads`. Returns `[unreads channels]`

[unreads](#unreads-2) \| [channels](#channels-1)
### Pokes
#### %channel-action

Takes an [a-channels](#a-channels), which can be one of several actions. Each action is covered below.

##### %create
```hoon
=c -build-file /=groups=/sur/channel/hoon
=g -build-file /=groups=/sur/groups/hoon

=create-channel :*
  %chat
  %mychat
  [our %mygroup]
  title='My chat channel'
  description='Chat about things'
  readers=~
  writers=~
==

:channels &channel-action [%create create-channel]
```

Create a channel. This also updates the metadata in the %groups agent.

[create-channel](#create-channel)

##### %channel
[nest](#nest)

> ###### %join
> ```hoon
> =nest [%chat ~palnet-sampel %mychat]
> =action [%join group=[~sampel-palnet %mygroup]]
> 
> :channels &channel-action [%channel nest action]
> ```
> 
> Join a channel. In this example, ~sampel-palnet is the group host and ~palnet-sampel is the channel host.

Reviewer: Humbly requesting more obviously distinct sample `@p`s. Something like:

````suggestion
###### %join
```hoon
=nest [%chat ~sampel-banter %mychat]
=action [%join group=[~hostes %mygroup]]

:channels &channel-action [%channel nest action]
```

Join a channel. In this example, ~hostes is the group host and ~sampel-banter is the channel host.
````

###### %leave
```hoon
=nest [%chat ~palnet-sampel %mychat]
=action [%leave ~]

:channels &channel-action [%channel nest action]
```

Leave a channel
###### %read
```hoon
=nest [%chat our %mychat]
=action [%read ~]

:channels &channel-action [%channel nest action]
```

Mark a channel as read.
###### %read-at
```hoon
=nest [%chat our %mychat]
=action [%read-at now]

:channels &channel-action [%channel nest action]
```

Mark a channel as read up to a certain time.
###### %watch
```hoon
=nest [%chat our %mychat]
=action [%watch ~]

:channels &channel-action [%channel nest action]
```

Watch a channel for unreads.
###### %unwatch
```hoon
=nest [%chat our %mychat]
=action [%unwatch ~]

:channels &channel-action [%channel nest action]
```

> Stop watching a channel for unreads.

Reviewer: What does it mean to "watch a channel for unreads"?

##### Actions

> All of the previous pokes affected your local ship (joining and leaving, unreads, and so on). The following pokes are *actions*, meaning they will be synced across the network. When `%channels` on your ship hears an action, it issues a *command* to the host's `%channels-server` (see the above [diagram](#diagram)).

Reviewer: Should probably explain this ahead of time. But also, the previous ones were actions, too, by our type naming convention. Marking some as "affects local state only" is fair, but shouldn't be made to seem like a big deal, and certainly shouldn't conflate terminology (especially if we're not talking about `%channels-server`/`$command`s here).

The heading for this could be nicer, but per the above, I'd argue this doesn't need a separate heading/nesting level at all.

Reviewer: Looking at the `/sur` file, I can see how this emphasis came about: the post & reply action types are defined in terms of their command types. Considering that the action types are the outward-facing ones, that should probably be the other way around. I may touch that up.

Author: @arthyn what do you think?

:channels &channel-action [%channel nest action]
```

Add a reaction to a reply

[id-post](#id-post) \| [id-reply](#id-reply) \| [react](#react)
###### %post %reply %del-react
```hoon
=nest [%chat our %mychat]
=reply-action [%del-react ~2023.2.2 our]
=post-action [%reply ~2023.1.1 reply-action]
=action [%post post-action]
:channels &channel-action [%channel nest action]
```

Delete a reaction from a reply

[id-post](#id-post) | [id-reply](#id-reply)
###### %view
```hoon
=nest [%chat our %mychat]
=action [%view %grid]
:channels &channel-action [%channel nest action]
```

Set the display format for a channel

[view](#view-1)
###### %sort
```hoon
=nest [%chat our %mychat]
=action [%sort %alpha]
:channels &channel-action [%channel nest action]
```

Set the sorting mechanism for a channel

[sort](#sort-1)
###### %set-order
```hoon
=nest [%chat our %mychat]
=action [%set-order `~[~2023.1.1 ~2023.2.2]]
:channels &channel-action [%channel nest action]
```

Set manual post ordering for a channel

[arranged-posts](#arranged-posts)
###### %add-writers
```hoon
=nest [%chat our %mychat]
=action [%add-writers (sy ~[%role1 %role2])]
:channels &channel-action [%channel nest action]
```

> Enable writing to a channel for a certain set of roles

Reviewer: Here and below, "allow" and "disallow" might be better wording than "enable" and "disable" respectively. More evocative of dealing with permissions.  
Maybe also just "certain roles".

[sect:groups](#sectgroups)

###### %del-writers
```hoon
=nest [%chat our %mychat]
=action [%del-writers (sy ~[%role1 %role2])]

:channels &channel-action [channel nest action]
```

Disable writing to a diary for a certain set of roles

[sect:groups](#sectgroups)

#### %channel-migration
Used internally to handle migrating from the previous version's state

#### Create a Group
```hoon
=g -build-file /=groups=/sur/groups/hoon

=create :*
  name=%mygroup
  title='My group'
  description='This is my group'
  image=''
  cover=''
  cordon=`cordon:g`[%open ban=[ships=~ ranks=~]]
  members=`(jug ship sect:g)`(~(put ju *(jug ship sect:g)) our %admin)
  secret=%.n
  ==

:groups &group-create create
```

This is done with `%groups` instead of `%channels`, but it's included here since it's necessary for testing.
### Subscriptions
#### /unreads

> Subscribe to unread & preview information. Each fact is a `[channels unreads]`

Reviewer: These should all be telling the mark for the facts, in addition to the type contained within that mark.

[channels](#channels) \| [unreads](#unreads)
#### /

> "Firehose" subscription path. Each fact is an `r-channels` (response channels)

Reviewer: I'm not sure if "firehose" is a broadly understood technical term. Consider elaborating on what exactly you mean by it.

[r-channels](#r-channels)

> #### /[kind]/[ship]/[name]
> (This is effectively /[nest])
> 
> Similar to `/`, but only include updates for a particular channels. Each fact is an `r-channels` (response channels)
> 
> [nest](#nest) \| [r-channels](#r-channels)

Reviewer: `/[nest]`, especially written next to all these other examples, suggests a path with a single element in it. If you want to point this out, I would just write something like:

```suggestion
#### /[kind]/[ship]/[name]

Similar to `/`, but only include updates for a particular channels. Note how the path fully describes a specific [`$nest`](#nest). Each fact is an [`$r-channels`](#r-channels) (channels response).
```

For other paths, I would either repeat the same short phrasing, or omit it entirely.

#### /said/[kind]/[ship]/[name]/post/[time]/[(unit reply)]
(This is effectively /said/[nest]/post/[time]/[(unit reply)])

Read a reference. Facts are either of the `%channel-denied` mark (meaning you cannot view the channel) or of the `%channel-said` mark and the `said` type

[nest](#nest) \| [said](#said)

> ## Types

Reviewer: Unclear to me if we should include these here, or just do permalinks to the type definitions in `/sur/channel`. Arguably any comments on the types we give here should be included in the sur file itself.

I guess having them in the document with anchor tags makes it faster for the reader to browser-back-button (#bbb) between the definitions and the part they were reading...

### r-channel
```hoon
+$  r-channel
  $%  [%posts =kind-data-posts]
      [%post id=id-post =r-post]
      [%set-order order=pin-channel-posts]
      [%view =view]
      [%sort =sort]
      [%perm =perm]
    ::
      [%create =perm]
      [%join group=flag:g]
      [%leave ~]
    ::
      [%read ~]
      [%read-at =time]
      [%watch ~]
      [%unwatch ~]
  ==
```

Response (subscriber to client communication) for a channel.

### channels
```hoon
+$  channels  (map nest channel)
```

A map of [nest](#nest) to [channel](#channel)

### channel
```hoon
++  channel
  |^  ,[global local]
  +$  global
    $:  posts=posts
        order=arranged-posts
        =view
        =sort
        =perm
    ==
  ::
  +$  local
    $:  =net
        =remark
    ==
  --
```

Posts and view/sort/order info for a channel

[posts](#posts) \| [arranged-posts](#arranged-posts) \| [view](#view) \| [sort](#sort) \| [perm](#perm)

### perm
```hoon
+$  perm
  $:  writers=(set sect:g)
      group=flag:g
  ==
```

Permissions for a channel

[sect:groups](#sectgroups) \| [flag:groups](#flaggroups)
### v-channel
Internal representation of a channel. Contains syncing information and revision numbers.

> [channel](#channel)

Reviewer: This one is never referenced, and described as "internal", so per my other comments we shouldn't be including this.

### nest
```hoon
+$  nest  [=kind =ship name=term]
```

Uniquely identifies a channel.

[kind](#kind)

### kind
```hoon
+$  kind  ?(%notebook %gallery %chat)
```

> Channel type.

Reviewer: I would probably put these at the very top of the types section, and generally order them so that primitives/foundationals come first. Right now the first type is described as "A nest and a note", but we haven't learned what either of those are.
